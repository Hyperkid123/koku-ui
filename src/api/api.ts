import axios, { AxiosRequestConfig } from 'axios';

export interface PagedResponse<D = any> {
  count: number;
  previous: string;
  next: string;
  results: D[];
}

export function initApi({ version }: { version: string }) {
  axios.defaults.baseURL = `/r/insights/platform/cost-management/api/${version}/`;
  axios.interceptors.request.use(authInterceptor);
}

export function authInterceptor(
  reqConfig: AxiosRequestConfig
): AxiosRequestConfig {
  return {
    ...reqConfig,
    headers: {
      ...reqConfig.headers,
    },
  };
}
