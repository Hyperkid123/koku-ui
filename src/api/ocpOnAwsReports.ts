import axios from 'axios';
import { Omit } from 'react-redux';
import { AwsReportData } from './awsReports';
import { OcpDatum } from './ocpReports';

export interface OcpOnAwsDatum {
  value: number;
  units: string;
}

// Todo: Remove capacity, limit, & request?
export interface OcpOnAwsReportValue {
  account?: string;
  account_alias?: string;
  capacity?: OcpOnAwsDatum;
  cost?: OcpOnAwsDatum;
  cluster?: string;
  cluster_alias?: string;
  count?: OcpOnAwsDatum;
  date: string;
  delta_percent?: number;
  delta_value?: number;
  derived_cost: OcpDatum;
  infrastructure_cost: OcpDatum;
  limit?: OcpOnAwsDatum;
  node?: string;
  project?: string;
  region?: string;
  request?: OcpOnAwsDatum;
  service?: string;
  usage: OcpOnAwsDatum;
}

export interface GroupByAccountData extends Omit<AwsReportData, 'accounts'> {
  account: string;
}

export interface GroupByClusterData
  extends Omit<OcpOnAwsReportData, 'clusters'> {
  service: string;
}

export interface GroupByNodeData extends Omit<OcpOnAwsReportData, 'nodes'> {
  region: string;
}

export interface GroupByProjectData
  extends Omit<OcpOnAwsReportData, 'projects'> {
  account: string;
}

export interface GroupByRegionData extends Omit<AwsReportData, 'regions'> {
  region: string;
}

export interface GroupByServiceData extends Omit<AwsReportData, 'services'> {
  service: string;
}

export interface OcpOnAwsReportData {
  accounts?: GroupByAccountData[];
  clusters?: GroupByClusterData[];
  date?: string;
  delta_percent?: number;
  delta_value?: number;
  nodes?: GroupByNodeData[];
  projects?: GroupByProjectData[];
  regions?: GroupByRegionData[];
  services?: GroupByServiceData[];
  values?: OcpOnAwsReportValue[];
}

export interface OcpOnAwsReportMeta {
  delta?: {
    percent: number;
    value: number;
  };
  group_by?: {
    [group: string]: string[];
  };
  order_by?: {
    [order: string]: string;
  };
  filter?: {
    [filter: string]: any;
  };
  total?: {
    capacity?: OcpOnAwsDatum;
    cost: OcpOnAwsDatum;
    derived_cost: OcpOnAwsDatum;
    infrastructure_cost: OcpOnAwsDatum;
    limit?: OcpOnAwsDatum;
    request?: OcpOnAwsDatum;
    usage?: OcpOnAwsDatum;
  };
  count: number;
}

export interface OcpOnAwsReportLinks {
  first: string;
  previous?: string;
  next?: string;
  last: string;
}

export interface OcpOnAwsReport {
  meta: OcpOnAwsReportMeta;
  links: OcpOnAwsReportLinks;
  data: OcpOnAwsReportData[];
}

export const enum OcpOnAwsReportType {
  cost = 'cost',
  cpu = 'cpu',
  database = 'database',
  instanceType = 'instance_type',
  memory = 'memory',
  network = 'network',
  storage = 'storage',
  tag = 'tag',
  volume = 'volume',
}

export const ocpOnAwsReportTypePaths: Record<OcpOnAwsReportType, string> = {
  [OcpOnAwsReportType.cost]: 'reports/openshift/infrastructures/aws/costs/',
  [OcpOnAwsReportType.cpu]: 'reports/openshift/compute/',
  [OcpOnAwsReportType.database]: 'reports/openshift/infrastructures/aws/costs/',
  [OcpOnAwsReportType.instanceType]:
    'reports/openshift/infrastructures/aws/instance-types/',
  [OcpOnAwsReportType.memory]: 'reports/openshift/memory/',
  [OcpOnAwsReportType.network]: 'reports/openshift/infrastructures/aws/costs/',
  [OcpOnAwsReportType.storage]:
    'reports/openshift/infrastructures/aws/storage/',
  [OcpOnAwsReportType.tag]: 'tags/openshift/',
  [OcpOnAwsReportType.volume]: 'reports/openshift/volumes/',
};

export function runReport(reportType: OcpOnAwsReportType, query: string) {
  const path = ocpOnAwsReportTypePaths[reportType];
  const insights = (window as any).insights;
  if (
    insights &&
    insights.chrome &&
    insights.chrome.auth &&
    insights.chrome.auth.getUser
  ) {
    return insights.chrome.auth.getUser().then(() => {
      return axios.get<OcpOnAwsReport>(`${path}?${query}`);
    });
  } else {
    return axios.get<OcpOnAwsReport>(`${path}?${query}`);
  }
}
