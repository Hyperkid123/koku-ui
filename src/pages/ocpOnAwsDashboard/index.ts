import { hot } from 'react-hot-loader';
import OcpOnAwsDashboard from './ocpOnAwsDashboard';

export default hot(module)(OcpOnAwsDashboard);
