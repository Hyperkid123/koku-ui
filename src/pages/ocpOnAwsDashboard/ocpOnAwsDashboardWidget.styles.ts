import { StyleSheet } from '@patternfly/react-styles';
import { global_spacer_xl } from '@patternfly/react-tokens';

export const styles = StyleSheet.create({
  tabItems: {
    marginTop: global_spacer_xl.value,
  },
});
