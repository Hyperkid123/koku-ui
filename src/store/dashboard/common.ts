export interface ValueFormatOptions {
  fractionDigits?: number;
}

export interface DashboardWidget {
  availableTabs?: any[];
  currentTab: any;
  details: {
    costKey?: string /** i18n label key */;
    formatOptions: ValueFormatOptions;
    showUnits?: boolean;
    showUsageLegendLabel?: boolean;
    usageFormatOptions?: ValueFormatOptions;
    usageKey?: string /** i18n label key */;
  };
  filter?: {
    limit?: number;
    service?: string;
  };
  id: number;
  isDetailsLink?: boolean;
  isHorizontal?: boolean;
  isUsageFirst?: boolean;
  reportType: any;
  /** i18n key for the title. passed { startDate, endDate, month, time } */
  titleKey: string;
  tabsFilter?: {
    limit?: number;
    service?: string;
  };
  trend: {
    titleKey: string;
    type: number;
    formatOptions: ValueFormatOptions;
  };
  topItems: {
    formatOptions: {};
  };
}
