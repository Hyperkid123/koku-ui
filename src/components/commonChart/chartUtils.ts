import { AwsReport } from 'api/awsReports';
import { OcpReport } from 'api/ocpReports';
import format from 'date-fns/format';
import getDate from 'date-fns/get_date';
import startOfMonth from 'date-fns/start_of_month';
import { FormatOptions, ValueFormatter } from 'utils/formatValue';
import {
  ComputedAwsReportItem,
  getComputedAwsReportItems,
  getIdKeyForGroupBy,
} from 'utils/getComputedAwsReportItems';
import {
  ComputedOcpReportItem,
  getComputedOcpReportItems,
} from 'utils/getComputedOcpReportItems';
import { SortDirection } from 'utils/sort';

export interface ChartDatum {
  x: string | number;
  y: number;
  key: string | number;
  name?: string | number;
  units: string;
}

export const enum ChartType {
  rolling,
  daily,
  monthly,
}

export function isAwsReport(
  report: AwsReport | OcpReport
): report is AwsReport {
  const groupById = report ? getIdKeyForGroupBy(report.group_by) : 'date'; // default key
  return groupById !== 'date';
}

export function transformReport(
  report: AwsReport | OcpReport,
  type: ChartType = ChartType.daily,
  key: any = 'date'
): ChartDatum[] {
  if (!report) {
    return [];
  }
  const items = {
    report,
    idKey: key,
    sortKey: 'id',
    sortDirection: SortDirection.desc,
  } as any;
  const computedItems = isAwsReport(report)
    ? getComputedAwsReportItems(items)
    : getComputedOcpReportItems(items);
  if (type === ChartType.daily) {
    return computedItems.map(i => createDatum(i.total, i, key));
  }

  if (type === ChartType.monthly) {
    return computedItems.map(i => createDatum(i.total, i, key));
  }

  return computedItems.reduce<ChartDatum[]>((acc, d) => {
    const prevValue = acc.length ? acc[acc.length - 1].y : 0;
    return [...acc, createDatum(prevValue + d.total, d, key)];
  }, []);
}

export function createDatum(
  value: number,
  computedItem: ComputedAwsReportItem | ComputedOcpReportItem,
  idKey = 'date'
): ChartDatum {
  const xVal = idKey === 'date' ? getDate(computedItem.id) : computedItem.label;
  return {
    x: xVal,
    y: parseFloat(value.toFixed(2)),
    key: computedItem.id,
    name: computedItem.id,
    units: computedItem.units,
  };
}

export function getDatumDateRange(datums: ChartDatum[]): [Date, Date] {
  if (!datums.length) {
    const today = new Date();
    const firstOfMonth = startOfMonth(today);
    return [firstOfMonth, today];
  }

  const start = new Date(datums[0].key + 'T00:00:00');
  const end = new Date(datums[datums.length - 1].key + 'T00:00:00');
  return [start, end];
}

export function getDateRangeString(datums: ChartDatum[]) {
  const [start, end] = getDatumDateRange(datums);
  const monthName = format(start, 'MMM');
  const startDate = getDate(start);
  const endDate = getDate(end);
  return `${monthName} ${getDate(start)}${
    startDate !== endDate ? ` - ${endDate}` : ''
  }`;
}

export function getTooltipLabel(
  datum: ChartDatum,
  formatValue: ValueFormatter,
  formatOptions?: FormatOptions,
  idKey: any = 'date'
) {
  if (!datum.key) {
    return '';
  }
  if (idKey === 'date') {
    const date = format(datum.key, 'MMM D YYYY');
    return `${date}: ${formatValue(datum.y, datum.units, formatOptions)}`;
  }
  return datum.key.toString();
}
