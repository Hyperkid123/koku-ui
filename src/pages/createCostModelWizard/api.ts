import { fetchCostModels } from 'api/costModels';
import { fetchProviders } from 'api/providers';

export const fetchSources = ({ type, page, perPage, query }) => {
  const offset = (page - 1) * perPage;
  const limit = perPage;
  const queryParam = Object.keys(query).reduce(
    (acc, cur) =>
      acc ? `${acc}&${cur}=${query[cur]}` : `${cur}=${query[cur]}`,
    ''
  );
  return fetchProviders(
    `type=${type}&limit=${limit}&offset=${offset}&${queryParam}`
  )
    .then(sources => sources.data.data)
    .then(sources => {
      return fetchCostModels().then(cms => ({
        costmodels: cms.data.data,
        sources,
      }));
    })
    .then(({ costmodels, sources }) => {
      const cmsHash = costmodels.reduce((acc, curr) => {
        curr.provider_uuids.forEach(provider_uuid => {
          acc[provider_uuid] = curr.name;
        });
        return acc;
      }, {});
      return sources.map(src => ({
        name: src.name,
        costmodel: cmsHash[src.uuid],
        uuid: src.uuid,
        selected: false,
      }));
    });
};
