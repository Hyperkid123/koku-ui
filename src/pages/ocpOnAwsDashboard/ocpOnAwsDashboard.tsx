import { Grid, GridItem } from '@patternfly/react-core';
import React from 'react';
import { InjectedTranslateProps, translate } from 'react-i18next';
import { connect } from 'react-redux';
import { createMapStateToProps } from 'store/common';
import { ocpOnAwsDashboardSelectors } from 'store/ocpOnAwsDashboard';
import { OcpOnAwsDashboardWidget } from './ocpOnAwsDashboardWidget';

type OcpOnAwsDashboardOwnProps = InjectedTranslateProps;

interface OcpOnAwsDashboardStateProps {
  widgets: number[];
}

interface OcpOnAwsDashboardDispatchProps {
  selectWidgets?: typeof ocpOnAwsDashboardSelectors.selectWidgets;
}

type OcpOnAwsDashboardProps = OcpOnAwsDashboardOwnProps &
  OcpOnAwsDashboardStateProps &
  OcpOnAwsDashboardDispatchProps;

const OcpOnAwsDashboardBase: React.SFC<OcpOnAwsDashboardProps> = ({
  selectWidgets,
  t,
  widgets,
}) => (
  <div>
    <Grid gutter="md">
      {widgets.map(widgetId => {
        const widget = selectWidgets[widgetId];
        return Boolean(widget.isHorizontal) ? (
          <GridItem sm={12} key={widgetId}>
            <OcpOnAwsDashboardWidget widgetId={widgetId} />
          </GridItem>
        ) : (
          <GridItem xl={4} lg={6} key={widgetId}>
            <OcpOnAwsDashboardWidget widgetId={widgetId} />
          </GridItem>
        );
      })}
    </Grid>
  </div>
);

const mapStateToProps = createMapStateToProps<
  OcpOnAwsDashboardOwnProps,
  OcpOnAwsDashboardStateProps
>(state => {
  return {
    selectWidgets: ocpOnAwsDashboardSelectors.selectWidgets(state),
    widgets: ocpOnAwsDashboardSelectors.selectCurrentWidgets(state),
  };
});

const OcpOnAwsDashboard = translate()(
  connect(
    mapStateToProps,
    {}
  )(OcpOnAwsDashboardBase)
);

export default OcpOnAwsDashboard;
