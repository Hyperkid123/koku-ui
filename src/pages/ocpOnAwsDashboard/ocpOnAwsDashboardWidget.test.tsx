jest
  .mock('date-fns/start_of_month')
  .mock('date-fns/get_date')
  .mock('date-fns/format')
  .mock('date-fns/get_month');

import { OcpOnAwsReportType } from 'api/ocpOnAwsReports';
import { ChartType } from 'components/charts/commonChart/chartUtils';
import formatDate from 'date-fns/format';
import getDate from 'date-fns/get_date';
import getMonth from 'date-fns/get_month';
import startOfMonth from 'date-fns/start_of_month';
import { shallow } from 'enzyme';
import React from 'react';
import { FetchStatus } from 'store/common';
import { OcpOnAwsDashboardTab } from 'store/ocpOnAwsDashboard';
import { mockDate } from 'testUtils';
import {
  getIdKeyForTab,
  OcpOnAwsDashboardWidgetBase,
  OcpOnAwsDashboardWidgetProps,
} from './ocpOnAwsDashboardWidget';

const props: OcpOnAwsDashboardWidgetProps = {
  widgetId: 1,
  id: 1,
  t: jest.fn(v => v),
  current: { data: [] },
  previous: { data: [] },
  tabs: { data: [] },
  fetchReports: jest.fn(),
  updateTab: jest.fn(),
  titleKey: 'title',
  reportType: OcpOnAwsReportType.cost,
  trend: {
    type: ChartType.rolling,
    titleKey: 'trend title',
    formatOptions: {},
  },
  status: FetchStatus.none,
  currentQuery: '',
  previousQuery: '',
  tabsQuery: '',
  details: {
    labelKey: 'detail label',
    labelKeyContext: 'label context',
    descriptionKeyRange: 'detail description range',
    descriptionKeySingle: 'detail description single',
    formatOptions: {},
  },
  topItems: { formatOptions: {} },
  availableTabs: [OcpOnAwsDashboardTab.projects],
  currentTab: OcpOnAwsDashboardTab.projects,
};

const getDateMock = getDate as jest.Mock;
const formatDateMock = formatDate as jest.Mock;
const startOfMonthMock = startOfMonth as jest.Mock;
const getMonthMock = getMonth as jest.Mock;

beforeEach(() => {
  mockDate();
  getDateMock.mockReturnValue(1);
  formatDateMock.mockReturnValue('formated date');
  startOfMonthMock.mockReturnValue(1);
  getMonthMock.mockReturnValue(1);
});

test('reports are fetched on mount', () => {
  shallow(<OcpOnAwsDashboardWidgetBase {...props} />);
  expect(props.fetchReports).toBeCalledWith(props.widgetId);
});

test('title is translated', () => {
  shallow(<OcpOnAwsDashboardWidgetBase {...props} />);
  expect(getTranslateCallForKey(props.titleKey)).toMatchSnapshot();
});

test('detail label is translated', () => {
  shallow(<OcpOnAwsDashboardWidgetBase {...props} />);
  expect(getTranslateCallForKey(props.details.unitsKey)).toMatchSnapshot();
});

test('subtitle is translated with single date', () => {
  shallow(<OcpOnAwsDashboardWidgetBase {...props} />);
  expect(
    getTranslateCallForKey('ocp_on_aws_dashboard.widget_subtitle')
  ).toMatchSnapshot();
});

test('subtitle is translated with date range', () => {
  getDateMock.mockReturnValueOnce(2);
  shallow(<OcpOnAwsDashboardWidgetBase {...props} />);
  expect(
    getTranslateCallForKey('ocp_on_aws_dashboard.widget_subtitle')
  ).toMatchSnapshot();
});

test('trend title is translated', () => {
  shallow(<OcpOnAwsDashboardWidgetBase {...props} />);
  expect(getTranslateCallForKey(props.trend.titleKey)).toMatchSnapshot();
});

test('id key for dashboard tab is the tab name in singular form', () => {
  [
    OcpOnAwsDashboardTab.clusters,
    OcpOnAwsDashboardTab.nodes,
    OcpOnAwsDashboardTab.projects,
  ].forEach(value => {
    expect(getIdKeyForTab(value)).toEqual(value.slice(0, -1));
  });

  expect(getIdKeyForTab(OcpOnAwsDashboardTab.projects)).toEqual('project');
});

function getTranslateCallForKey(key: string) {
  return (props.t as jest.Mock).mock.calls.find(([k]) => k === key);
}
