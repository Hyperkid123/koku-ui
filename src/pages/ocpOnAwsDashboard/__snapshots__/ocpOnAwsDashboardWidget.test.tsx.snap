// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`detail label is translated 1`] = `undefined`;

exports[`subtitle is translated with date range 1`] = `
Array [
  "ocp_on_aws_dashboard.widget_subtitle",
  Object {
    "month": 1,
  },
]
`;

exports[`subtitle is translated with single date 1`] = `
Array [
  "ocp_on_aws_dashboard.widget_subtitle",
  Object {
    "month": 1,
  },
]
`;

exports[`title is translated 1`] = `
Array [
  "title",
  Object {
    "endDate": "formated date",
    "month": 1,
    "startDate": "formated date",
  },
]
`;

exports[`trend title is translated 1`] = `
Array [
  "trend title",
]
`;
